File: linux/samples/enc_mp4_avc_aac_push/avencode.cpp
#include <iostream>
#include <fstream>
#include <stdexcept>

// memmove in AVBlocks
#include <string.h>
#include <AVBlocks.h>
#include <PrimoReference++.h>
#include <PrimoUString.h>

#include "util.h"

using namespace primo::codecs;
using namespace primo::avblocks;
using namespace primo::error;
using namespace std;

inline void printError(const char* action, const ErrorInfo* e)
{
    if (action)
    {
            std::cout << action << ": ";
    }

    if (ErrorFacility::Success == e->facility())
    {
            std::cout << "Success" << std::endl;
            return;
    }

    if (e->message())
    {
            std::cout << primo::ustring(e->message()) << " ";
    }

    std::cout << "facility:" << e->facility() << " error:" << e->code() << "" << std::endl;
}

class UncompressedAVSplitter
{
public:

	~UncompressedAVSplitter()
	{
		inframe->release();
	}

	UncompressedAVSplitter() :
		inframe(primo::avblocks::Library::createMediaSample()),
		uncompressed_frame_size(0),
		inframe_count(0),
		eos(true),
		media_type(MediaType::Unknown)
	{}

	void init(const MediaSocket* socket, const string& filename)
	{
		if (!socket || filename.empty()) {
			return;
		}

		infile.open(filename.c_str(), ios::in | ios::binary);
		if (!infile)
		{
			throw std::logic_error("Cannot open file");
		}

		StreamInfo* sinfo = socket->pins()->at(0)->streamInfo();

		media_type = sinfo->mediaType();
		
		if (sinfo->mediaType() == MediaType::Video) 
		{
			init((VideoStreamInfo*)sinfo);
		}
		else if (sinfo->mediaType() == MediaType::Audio) 
		{
			init((AudioStreamInfo*)sinfo);
		}
		else 
		{
			throw std::logic_error("Unsupported: unknown media type");
		}

		eos = false;
	}

	MediaSample* get_frame()
	{
		if (eos)
			return NULL;

		if (media_type == MediaType::Video)
		{
			infile.read((char*)inframe->buffer()->start(), uncompressed_frame_size);
			int bytes_read = (int)infile.gcount();

			if (bytes_read < uncompressed_frame_size) {
				eos = true;
			}
			else {
				inframe->buffer()->setData(0, uncompressed_frame_size);
				double time = inframe_count / frame_rate;
				inframe->setStartTime(time);
				inframe->setEndTime(-1.0);
				++inframe_count;
			}
		}
		else if (media_type == MediaType::Audio)
		{
			if (inframe->buffer()->dataSize() == 0) {
				infile.read((char*)inframe->buffer()->start(), uncompressed_frame_size);
				int bytes_read = (int)infile.gcount();

				if (0 == bytes_read) {
					eos = true;
				}
				else {
					inframe->buffer()->setData(0, bytes_read);
					double time = inframe_count / frame_rate;
					inframe->setStartTime(time);
					inframe->setEndTime(-1.0);
					++inframe_count;
				}
			}
		}

		return (eos ? NULL : inframe);
	}

	private:
	
	void init(VideoStreamInfo* vinfo)
	{
		if (vinfo->streamType() == StreamType::UncompressedVideo)
		{
			uncompressed_frame_size = vinfo->frameWidth() * vinfo->frameHeight() * 3 / 2; // assume YUV 4:2:0
			MediaBuffer* buffer = (primo::avblocks::Library::createMediaBuffer(uncompressed_frame_size));
			inframe->setBuffer(buffer);
			buffer->release();
		}
		else
		{
			throw std::logic_error("Unsupported: video input is compressed");
		}

		frame_rate = vinfo->frameRate();
	}

	void init(AudioStreamInfo* ainfo)
	{
		if (ainfo->streamType() == StreamType::LPCM)
		{
			if (ainfo->bytesPerFrame() == 0) {
				ainfo->setBytesPerFrame(ainfo->bitsPerSample() / 8 * ainfo->channels());
			}
			
			frame_rate = 10;
			uncompressed_frame_size = ainfo->bytesPerFrame() * ainfo->sampleRate() / (int32_t)frame_rate;
			MediaBuffer* buffer(primo::avblocks::Library::createMediaBuffer(uncompressed_frame_size));
			inframe->setBuffer(buffer);
			buffer->release();
		}
		else
		{
			throw std::logic_error ("Unsupported: audio input is compressed");
		}
	}

	MediaType::Enum media_type;
	MediaSample* inframe;
	std::ifstream infile;
	int uncompressed_frame_size;
	int inframe_count;
	double frame_rate;
	bool eos;
};

struct TrackState
{
	enum 
	{
		Disabled = -1,
	};

	TrackState() : frame(NULL), index(Disabled), frame_count(0), progress (-1.0) {}

	MediaSample* frame;
	int32_t index;
	int32_t frame_count;
	double progress;
};

TrackState* select_mux_track(TrackState& vtrack, TrackState& atrack)
{
	if (vtrack.index != TrackState::Disabled && atrack.index != TrackState::Disabled)
	{
		if (!vtrack.frame)
			return &vtrack;

		if (!atrack.frame)
			return &atrack;

		return vtrack.frame->startTime() <= atrack.frame->startTime() ? &vtrack : &atrack;
	}
	
	if (vtrack.index != TrackState::Disabled) {
		return &vtrack;
	}

	if (atrack.index != TrackState::Disabled) {
		return &atrack;
	}

	return NULL;
}


/*
mix and encode uncompressed audio and video input files
*/
bool av_encode(MediaSocket* vinput, const string& vfile, 
				MediaSocket* ainput, const string& afile, 
				MediaSocket* output)
{
	bool_t res;

	TrackState vtrack, atrack;
	
	UncompressedAVSplitter vsplit, asplit;
	
	try {
		if (vinput) {
			cout << "video input file: \"" << vfile << "\"" << endl;
			vsplit.init(vinput, vfile);
			cout << "OK" << endl;
		}
		if (ainput) {
			cout << "audio input file: \"" << afile << "\"" << endl;
			asplit.init(ainput, afile);
			cout << "OK" << endl;
		}
	}
	catch (std::exception& ex) {
		cout << ex.what() << endl;
		return false;
	}

	// setup transcoder
	auto transcoder = primo::make_ref(primo::avblocks::Library::createTranscoder());
	// in order to use the OEM release for testing (without a valid license) the demo mode must be enabled.
	transcoder->setAllowDemoMode(1);
	
	int track_index = 0; // start index

	if (vinput) {
		transcoder->inputs()->add(vinput);
		vtrack.index = track_index++;
	}

	if (ainput) {
		transcoder->inputs()->add(ainput);
		atrack.index = track_index++;
	}

	transcoder->outputs()->add(output);
	res = transcoder->open();
	printError("transcoder open", transcoder->error());
	if (!res)
		return false;

	// transcoding loop
	for(;;)
	{
		if (vtrack.index != TrackState::Disabled && !vtrack.frame) {
			vtrack.frame = vsplit.get_frame();
		}
		
		if (atrack.index != TrackState::Disabled && !atrack.frame) {
			atrack.frame = asplit.get_frame();
		}

		TrackState* track = select_mux_track(vtrack, atrack);

		if (!track)
			break;

		// log
		if (track->frame) {
			if (track->frame->startTime() - track->progress >= 1.0) {
				track->progress = track->frame->startTime();
				cout << "track " << track->index << " frame #" << track->frame_count << " pts:" << track->frame->startTime() << endl;
			}
		}
		else {
			cout << "track " << track->index << " eos" << endl;
		}

		if (track->frame) {
			res = transcoder->push(track->index, track->frame);
			if (!res) {
				printError("transcoder push frame", transcoder->error());
				return false;
			}
			track->frame = NULL; // clear the muxed frame in order to read to the next one
			track->frame_count++;
		}
		else {
			res = transcoder->push(track->index, NULL);
			if (!res) {
				printError("transcoder push eos", transcoder->error());
				return false;
			}
			track->index = TrackState::Disabled; // disable track
		}
	}

	res = transcoder->flush();
	if (!res) {
		printError("transcoder flush", transcoder->error());
		return false;
	}

	transcoder->close();
        
        return true;
}

File: windows/net/samples/enc_mp4_avc_aac_push/AVEncode.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using PrimoSoftware.AVBlocks;

namespace EncMp4AvcAacPushSample
{
    class TrackState
    {
        public const int Disabled = -1;
        public MediaSample Frame;
        public int Index = Disabled;
        public int FrameCount = 0;
        public double Progress = -1.0;
    };

    public class UncompressedAVSplitter: IDisposable
    {
        private void Init(VideoStreamInfo vinfo)
        {
            if (vinfo.StreamType == StreamType.UncompressedVideo)
            {
                // YUV 420
                uncompressedFrameSize = vinfo.FrameWidth * vinfo.FrameHeight * 3 / 2;
                inframe.Buffer = new MediaBuffer(new byte[uncompressedFrameSize]);
                inframe.Buffer.SetData(0, 0); // needed because setting the buffer actually sets the data as well
            }
            else
            {
               throw new Exception("Unsupported: video input is compressed");
            }

            frameRate = vinfo.FrameRate;
        }

        private void Init(AudioStreamInfo ainfo)
        {
            if (ainfo.StreamType == StreamType.LPCM)
            {
                if (0 == ainfo.BytesPerFrame)
                {
                    ainfo.BytesPerFrame = ainfo.BitsPerSample / 8 * ainfo.Channels;
                }
                frameRate = 10;
                uncompressedFrameSize = ainfo.BytesPerFrame * ainfo.SampleRate / (int)frameRate;
                inframe.Buffer = new MediaBuffer(new byte[uncompressedFrameSize]);
                inframe.Buffer.SetData(0, 0); // needed because setting the buffer actually sets the data as well
            }
            else
            {
                throw new Exception("Unsupported: audio input is compressed");
            }
        }

        public void Init(MediaSocket socket, string filename)
        {
            if (socket == null || string.IsNullOrEmpty(filename))
            {
                return;
            }

            infile = new System.IO.FileStream(filename, FileMode.Open, FileAccess.Read);

            StreamInfo sinfo = socket.Pins[0].StreamInfo;
            mediaType = sinfo.MediaType;

            switch (mediaType)
            {
                case MediaType.Video:
                    Init(sinfo as VideoStreamInfo);
                    break;

                case MediaType.Audio:
                    Init(sinfo as AudioStreamInfo);
                    break;

                default:
                    throw new Exception("Unsupported: unknown media type");
            }

            eos = false;
        }

        public MediaSample GetFrame()
        {
            if (eos)
                return null;

            if (mediaType == MediaType.Video)
            {
                int bytesRead = infile.Read(inframe.Buffer.Start, 0, uncompressedFrameSize);

                if (bytesRead < uncompressedFrameSize)
                {
                    eos = true;
                }
                else
                {
                    inframe.Buffer.SetData(0, uncompressedFrameSize);
                    inframe.StartTime = inframeCount / frameRate;
                    inframe.EndTime = -1.0;
                    ++inframeCount;
                }
            }
            else if (mediaType == MediaType.Audio)
            {
                if (inframe.Buffer.DataSize == 0)
                {
                    int bytesRead = infile.Read(inframe.Buffer.Start, 0, uncompressedFrameSize);

                    if (0 == bytesRead)
                    {
                        eos = true;
                    }
                    else
                    {
                        inframe.Buffer.SetData(0, bytesRead);
                        inframe.StartTime = inframeCount / frameRate;
                        inframe.EndTime = -1;
                        ++inframeCount;
                    }
                }
            }

            return (eos ? null : inframe);
        }

        public void Dispose()
        {
            if (null != infile)
            {
                infile.Dispose();
            }
        }

        private MediaType mediaType = MediaType.Unknown;
        private MediaSample inframe = new MediaSample();
        private System.IO.FileStream infile;
        private int uncompressedFrameSize;
        private int inframeCount;
        private double frameRate;
        private bool eos = true;
    };

    public static class AVEncode
    {
        private static TrackState SelectMuxTrack(TrackState vtrack, TrackState atrack)
        {
            if (vtrack.Index != TrackState.Disabled && atrack.Index != TrackState.Disabled)
            {
                if (vtrack.Frame == null)
                    return vtrack;

                if (atrack.Frame == null)
                    return atrack;

                return vtrack.Frame.StartTime <= atrack.Frame.StartTime ? vtrack : atrack;
            }

            if (vtrack.Index != TrackState.Disabled)
                return vtrack;

            if (atrack.Index != TrackState.Disabled)
                return atrack;

            return null;
        }

        static void PrintError(string action, ErrorInfo e)
        {
            if (action != null)
            {
                Console.Write("{0}: ", action);
            }

            if (ErrorFacility.Success == e.Facility)
            {
                Console.WriteLine("Success");
                return;
            }
            else
            {
                Console.WriteLine("{0}, facility:{1} code:{2}", e.Message ?? "", e.Facility, e.Code);
            }
        }

        /*
        uncompressed audio and video input
        */
        public static bool Run(MediaSocket vinput, string vfile, MediaSocket ainput, string afile, MediaSocket output)
        {
            bool res;

            TrackState vtrack = new TrackState();
            TrackState atrack = new TrackState();

            using (UncompressedAVSplitter vsplit = new UncompressedAVSplitter())
            using (UncompressedAVSplitter asplit = new UncompressedAVSplitter())
            using (var transcoder = new Transcoder() { AllowDemoMode = true })
            {
                try
                {
                    if (vinput != null)
                    {
                        Console.WriteLine("video input file: \"{0}\"", vfile);
                        vsplit.Init(vinput, vfile);
                        Console.WriteLine("OK");
                    }

                    if (ainput != null)
                    {
                        Console.WriteLine("audio input file: \"{0}\"", afile);
                        asplit.Init(ainput, afile);
                        Console.WriteLine("OK");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.Message);
                    return false;
                }

                // setup transcoder
                int trackIndex = 0; // start index

                if (vinput != null)
                {
                    transcoder.Inputs.Add(vinput);
                    vtrack.Index = trackIndex++;
                }

                if (ainput != null)
                {
                    transcoder.Inputs.Add(ainput);
                    atrack.Index = trackIndex++;
                }

                transcoder.Outputs.Add(output);

                res = transcoder.Open();
                PrintError("transcoder open", transcoder.Error);
                if (!res)
                    return false;

                // transcoding loop
                for (;;)
                {
                    if (vtrack.Index != TrackState.Disabled && vtrack.Frame == null)
                    {
                        vtrack.Frame = vsplit.GetFrame();
                    }

                    if (atrack.Index != TrackState.Disabled && atrack.Frame == null)
                    {
                        atrack.Frame = asplit.GetFrame();
                    }

                    TrackState track = SelectMuxTrack(vtrack, atrack);

                    if (track == null)
                        break;

                    // log
                    if (track.Frame != null)
                    {
                        if (track.Frame.StartTime - track.Progress >= 1.0)
                        {
                            track.Progress = track.Frame.StartTime;
                            Console.WriteLine("track {0} frame #{1} pts:{2}", track.Index, track.FrameCount, track.Frame.StartTime);
                        }
                    }
                    else
                    {
                        Console.WriteLine("track {0} eos", track.Index);
                    }

                    if (track.Frame != null)
                    {
                        res = transcoder.Push(track.Index, track.Frame);
                        if (!res)
                        {
                            PrintError("transcoder push frame", transcoder.Error);
                            return false;
                        }
                        track.Frame = null; // clear the muxed frame in order to read to the next one
                        track.FrameCount++;
                    }
                    else
                    {
                        res = transcoder.Push(track.Index, null);
                        if (!res)
                        {
                            PrintError("transcoder push eos", transcoder.Error);
                            return false;
                        }
                        track.Index = TrackState.Disabled; // disable track
                    }
                }

                res = transcoder.Flush();
                if (!res)
                {
                    PrintError("transcoder flush", transcoder.Error);
                    return false;
                }
                transcoder.Close();
            }

            return true;
        }
    }
}
